Added notification threading.

brettp opened the pull request:
Fixes #409.

superalex commented on taiga/projects/notifications/services.py:
```diff
@@ -270,11 +273,24 @@ def send_sync_notifications(notification_id):
     template_name = _resolve_template_name(model, change_type=notification.history_type)
     email = _make_template_mail(template_name)
+    site = get_current()
+    domain = site.domain.split(":")[0] or site.domain
```
In other parts of the application we are reading site and domain from the app settings, settings.SITES["api"]["scheme"], settings.SITES["api"]["domain"]

superalex commented:
Besides my comment about the settings I really like the PR and think it's really useful! :+1: 

Could you please update the code so we can accept the PR?

superalex commented on taiga/projects/notifications/services.py:
```diff
@@ -270,11 +273,24 @@ def send_sync_notifications(notification_id):
+        id = obj.snapshot["ref"]
+    elif "slug" in obj.snapshot:
+        id = obj.snapshot["slug"]
+
```
I think a default allways existing value for id should be included

brettp commented:
@superalex: I found a problem with Outlook so would want to fix that first, too. I'll take care of the issues you mentioned while fixing that.

bameda commented on taiga/projects/notifications/services.py:
```diff
@@ -270,11 +273,24 @@ def send_sync_notifications(notification_id):
+               "References": "<%s/%s@%s>" % (notification.project.slug, id, domain),
+
+               "List-ID": 'Taiga/%s <taiga.%s@%s>'  % (notification.project.name, notification.project.slug, domain)}
 
```
Please, use str.format method instead of % ([PEP 3101](https://www.python.org/dev/peps/pep-3101/))

brettp commented:
I've updated the PR with the requested and changes, and attempted to fix the Outlook problems. Outlook doesn't use standard headers to thread messages, but I've added experimental support for their base64 encoded binary headers.

superalex commented:
It looks great but could you please add some tests for the new parts?

Regards!,

brettp commented:
@superalex I've added tests and cleaned up the MS Thread-ID function to make it less newbie-y. Let me know if you want to see any other changes!

superalex commented on tests/integration/test_notifications.py:
```diff
@@ -33,6 +37,7 @@
 from taiga.projects.userstories.serializers import UserStorySerializer
 from taiga.projects.tasks.serializers import TaskSerializer
 from taiga.permissions.permissions import MEMBERS_PERMISSIONS
+from taiga.projects.notifications import services as notification_services
```
This one is a duplicated, taiga.projects.notificationsservices is imported in line 31

superalex commented on taiga/projects/notifications/services.py:
```diff
@@ -356,3 +381,33 @@ def remove_watcher(obj, user):
+    import base64
+    import hashlib
+    import struct
+
```
Why not moving these imports to the top of the file?

superalex commented:
Thank you very much @brettp 

I was checking your PR with @bameda and we only have two minor comments :+1:

brettp commented:
@superalex and @bameda: Thanks! I've cleaned up the imports.

superalex commented:
Great!, thank you very much @brettp
